pkg/api/serialization_test.go

```diff
@@ -131,6 +132,22 @@ func roundTripSame(t *testing.T, group testapi.TestGroup, item runtime.Object, e
 	}
 }
 
+func TestSetControllerConversion(t *testing.T) {
+	rs := &extensions.ReplicaSet{}
+	gv := testapi.Extensions
+	fuzzInternalObject(t, gv.InternalGroupVersion(), rs, rand.Int63())
+
+	data, err := runtime.Encode(gv.Codec(), rs)
+	if err != nil {
+		t.Errorf("unexpected encoding error: %v", err)
+	}
+
+	rc := &v1.ReplicationController{}
+	if err := runtime.DecodeInto(testapi.Default.Codec(), data, rc); err != nil {
```

Comment:
@lavalamp I used the cross-group decoding helper you pointed me to but with no luck. 
```
serialization_test.go:153: unexpected error while enabling cross-group decoding: group "" is not a possible destination group in the given codec
```
I used both the legacy codec and versioned (`KUBE_TEST_API_TYPE=application/json`) and both testapi.Default and testapi.Extensions codec. Then I went ahead and created my own codec that knows how to do cross-group decoding but I still stumbled upon the panic above. 

I dug more and it seems that the codec never gets to decide about any cross-group conversion if you are trying to convert to a specific object `(into != nil)`.
```
k8s.io/kubernetes/pkg/runtime/serializer/versioning.(*codec).Decode(0xc8204c2a80, 0xc82000cd80, 0x46a, 0x46a, 0x0, 0x7ff6191c47f8, 0xc820344c60, 0x0, 0x0, 0xc82042b1c0, ...)
	/home/mkargaki/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/runtime/serializer/versioning/versioning.go:156 +0x3c8
k8s.io/kubernetes/pkg/runtime.DecodeInto(0x7ff6191d4998, 0xc8204c2a80, 0xc82000cd80, 0x46a, 0x46a, 0x7ff6191c47f8, 0xc820344c60, 0x0, 0x0)
	/home/mkargaki/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/runtime/codec.go:60 +0x98
k8s.io/kubernetes/pkg/api_test.TestSetControllerConversion(0xc8200154d0)
	/home/mkargaki/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/api/serialization_test.go:174 +0xf24
```
The method in question that switches to the convertor before looking into the decoding versions the codec supports:
```go
func (c *codec) Decode(data []byte, defaultGVK *unversioned.GroupVersionKind, into runtime.Object) (runtime.Object, *unversioned.GroupVersionKind, error) {
	versioned, isVersioned := into.(*runtime.VersionedObjects)
	if isVersioned {
		into = versioned.Last()
	}

       // uses the UniversalDeserializer if the test runs with KUBE_TEST_API_TYPE=application/json
	obj, gvk, err := c.decoder.Decode(data, defaultGVK, into)
	if err != nil {
		return nil, gvk, err
	}

	// if we specify a target, use generic conversion.
	if into != nil {
		if into == obj {
			if isVersioned {
				return versioned, gvk, nil
			}
			return into, gvk, nil
		}
		if err := c.convertor.Convert(obj, into); err != nil {   // <---- defers to the convertor
			return nil, gvk, err
		}
		if isVersioned {
			versioned.Objects = []runtime.Object{obj, into}
			return versioned, gvk, nil
		}
		return into, gvk, nil
	}

	// invoke a version conversion
	group := gvk.Group
	if defaultGVK != nil {
		group = defaultGVK.Group
	}
	var targetGV unversioned.GroupVersion
	if c.decodeVersion == nil {
		// convert to internal by default
		targetGV.Group = group
		targetGV.Version = runtime.APIVersionInternal
	} else {
		gv, ok := c.decodeVersion[group]  // <--- picks up the version to decode, converts later not included in this code snippet
		if !ok {
			// unknown objects are left in their original version
			if isVersioned {
				versioned.Objects = []runtime.Object{obj}
				return versioned, gvk, nil
			}
			return obj, gvk, nil
		}
		targetGV = gv
	}
// more
```